lgc/patch/PatchResourceCollect.cpp

```diff
@@ -1175,20 +1175,22 @@ void PatchResourceCollect::visitCallInst(CallInst &callInst) {
       // export call.
       m_deadCalls.push_back(&callInst);
 
-      InOutLocationInfo outLocInfo;
-      outLocInfo.setLocation(cast<ConstantInt>(callInst.getArgOperand(0))->getZExtValue());
-      outLocInfo.setComponent(cast<ConstantInt>(callInst.getArgOperand(1))->getZExtValue());
-      if (m_shaderStage == ShaderStageGeometry)
-        outLocInfo.setStreamId(cast<ConstantInt>(callInst.getArgOperand(2))->getZExtValue());
-      // Also, we remove the output location info from the map if it exists
-      auto &outLocInfoMap = m_resUsage->inOutUsage.outputLocInfoMap;
-      if (outLocInfoMap.count(outLocInfo) > 0)
-        outLocInfoMap.erase(outLocInfo);
-      // For GS, we remove transform feedback location info as well if it exists
-      if (m_shaderStage == ShaderStageGeometry) {
-        auto &locInfoXfbOutInfoMap = m_resUsage->inOutUsage.gs.locInfoXfbOutInfoMap;
-        if (locInfoXfbOutInfoMap.count(outLocInfo) > 0)
-          locInfoXfbOutInfoMap.erase(outLocInfo);
+      if (m_pipelineState->canPackOutput(m_shaderStage)) {
```

Comment:
Sorry. I didn't notice your discussion with Xue previously. This piece of code is added by me. I want to remove unused output components to save buffer/LDS store. And also reduce transform feedback outputs. This is for both packed pipeline and unpacked pipeline. But I didn't realize this will cause problems for unlinked pipeline. So please change the condition to `if (!unlinked)`. That is my real purpose.

Reply:
I think Xue's suggestion (that Steven has implemented) of using canPackOutput is better, because it allows for the case of compiling a part-pipeline, where you cannot pack the final-vertex-stage shader's outputs but you can pack other outputs (e.g. VS -> TCS).

Reply:
I agree with Tim.  Using "canPackOutput" seems to make the most sense.  Removing the location from the map is packing the output.  So this code should only be executed if you can pack the input.

Reply:
Thanks all for your reply. Let explain this issue:
In the function `visitCallInst`, we have a branch to remove undefined transform feedback values. Please see this:

```
    auto outputValue = callInst.getArgOperand(callInst.getNumArgOperands() - 1);
    if (isa<UndefValue>(outputValue)) {
      // NOTE: If an output value is undefined, we can safely drop it and remove the transform feedback output export
      // call.
      m_deadCalls.push_back(&callInst);
```

But we have already set up the map `locInfoXfbOutInfoMap` when builder is called. See `CreateWriteXfbOutput`. Therefore, we have to remove the in/out pack info from the maps `outputLocInfoMap` and `locInfoXfbOutInfoMap`.

Note that this is irrelevant to pack status. Even though the output is not packed, we still have to remove such undefined transform feedback values and their in/out location map items. Hence, if you use the condition `canPackOutput`, this piece of code will not be triggered. And in `PatchCopyShader::exportOutput`, when we traverse the map, the assert `assert(outputLocInfoMap.count(origLocInfo))` will be hit. This is why I suggest we use `unlinked` as the condition.

For future transform feedback implementation, we actually disable GS output packing. Therefore, all this piece of codes will be skipped entirely. And finally the assert in copy shader is hit.

Reply:
This sounds like something where should be a test for.  Can you create one?  That way I'll be able to see what is going on for myself.

Reply:
Please see this pipeline from the CTS: dEQP-VK.transform_feedback.fuzz.various_buffers.buffers110_instance_array_geometry

Some outputs are undefined values because they are not written by GS. To reproduce this issue, you must add those code to `InOutBuilder::CreateWriteXfbOutput`:

```
if (m_shaderStage == ShaderStageGeometry) {
    getPipelineState()->setPackOutput(ShaderStageGeometry, false);
    getPipelineState()->setPackInput(ShaderStageFragment, false);
   ...
}

This will cause the copy shader generation to go to unpacked path. If you use `canPackOutput` rather than `!isUnlinked`, the removal of map items will not be executed. And this will finally cause an assert in copy shader.
[xfb_fuzz.txt](https://github.com/GPUOpen-Drivers/llpc/files/6809782/xfb_fuzz.txt)

```

Reply:
I think I see the issue now.  Correct me if I am wrong.  "Packing" output is taking two or more output and merging them into a single output.  For example taking two vec2 outputs and outputting at a single vec4 output.  That is not what this code is doing.

I'll change this to check the next shader stage.  If the next stage is invalid, then we will not remove the entry in the map.  This should the partial-pipeline scenario that Tim mentioned to still work, and fix the problem that I am seeing.  Let me know if this is reasonable.

Also, we should have lit tests as much as possible.  They are much easier to run and debug than CTS.  The lit test can also be very specific with comments that explain what is suppose to happen if I were to do something that breaks it.  If something breaks in CTS, which in this case it will not because we need more changes to the repo, I do not have something guiding me on what is suppose to happen except the Vulkan SPEC.

Reply:
> Correct me if I am wrong. "Packing" output is taking two or more output and merging them into a single output. For example taking two vec2 outputs and outputting at a single vec4 output. That is not what this code is doing.

Yes, correct. This code is to clean up undefined outputs whether it is packed or unpacked. Because we remove the output call for those undefined values, we must remove them from the map items as well.

> I'll change this to check the next shader stage. If the next stage is invalid, then we will not remove the entry in the map. This should the partial-pipeline scenario that Tim mentioned to still work, and fix the problem that I am seeing. Let me know if this is reasonable.

This could be a solution.